Kuadrant/kuadrant-operator
# refactor: target status -> discoverability reconciler

# Description
Closes: https://github.com/Kuadrant/kuadrant-operator/issues/830 https://github.com/Kuadrant/kuadrant-operator/issues/829 https://github.com/Kuadrant/kuadrant-operator/issues/828 https://github.com/Kuadrant/kuadrant-operator/issues/827

* Refactors target status controller to seperate gateway and http route discoverability reconcilers
* Adds `xPolicyAffected` condition to Listeners status, listing accepted policies targeting the gateway and the listener (if any)
* Refactors `xPolicyAffected` condtions to list accepted policies in path that may affect the targetablle
  * Eg. Gateway condition will list only gateway policies, Listener status condition will list both gateway and listener policies, Route condition will list Gateway, Listener and route policies
 
# Verification
* Create cluster
```
make local-setup
```
* Create gateway RLP
```
kubectl apply -n gateway-system -f - <<EOF                                                                                                                     
apiVersion: kuadrant.io/v1beta3
kind: RateLimitPolicy
metadata:      
  name: gw-rlp           
spec:        
  targetRef:       
    group: gateway.networking.k8s.io
    kind: Gateway    
    name: kuadrant-ingressgateway
  overrides:      
    limits:         
      "global":   
        rates:   
        - limit: 5
          duration: 10
          unit: second        
        when:      
        - selector: source.address
          operator: neq  
          value: 127.0.0.1
    strategy: merge
EOF
```
* Verify Gateway `kuadrant.io/RateLimitPolicyAffected` condition is `Object affected by RateLimitPolicy [gateway-system/gw-rlp]` for `Gateway.Status.Condtions` and `Gateway.Status.Listener[].Condition` 
```
kubectl get gateway kuadrant-ingressgateway -n gateway-system -o yaml | yq '.status'
```
* Create HTTP Route
```
kubectl apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute 
metadata:
  name: toystore
spec:                               
  parentRefs:      
  - name: kuadrant-ingressgateway
    namespace: gateway-system
  hostnames:       
  - api.toystore.com  
  rules:       
  - matches:          
    - method: GET     
      path:                   
        type: PathPrefix  
        value: "/cars"            
    - method: GET      
      path:               
        type: PathPrefix
        value: "/dolls"                                                                                         
    backendRefs:                           
    - name: toystore
      port: 80
  - matches:
    - path:
        type: PathPrefix
        value: "/admin"
    backendRefs:
    - name: toystore
      port: 80
EOF
```
* Verify HTTP Route `kuadrant.io/RateLimitPolicyAffected` condition is `Object affected by RateLimitPolicy [gateway-system/gw-rlp]` for `Route.Status.Condtions` 
```
 kubectl get httproute toystore -o yaml | yq '.status'
```
* Creation section name rlp
```
kubectl apply -n gateway-system -f - <<EOF              
apiVersion: kuadrant.io/v1beta3
kind: RateLimitPolicy
metadata:
  name: section-rlp
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: kuadrant-ingressgateway
    sectionName: http
  limits:
    "specific":
      rates:
      - limit: 3
        duration: 5
        unit: second
      - limit: 20
        duration: 1
        unit: minute
EOF
```
* Verify the `gw-rlp` is listed along with `section-rlp`  for the targeted listener. All other listener status and the parent gatway condition should only list `gw-rlp`
```
kubectl get gateway kuadrant-ingressgateway -n gateway-system -o yaml | yq '.status'
```
* Verify http route condition also lists `gw-rlp` and `section-rlp` in the affection condition
```
 kubectl get httproute toystore -o yaml | yq '.status' 
```
* Create Route RLP
```
kubectl apply -f - <<EOF                 
apiVersion: kuadrant.io/v1beta3
kind: RateLimitPolicy
metadata:
  name: route-rlp
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: toystore
  limits:
    "specific":
      rates:
      - limit: 3
        duration: 5
        unit: second
      - limit: 20
        duration: 1
        unit: minute
EOF
```
* Verify http route condition also lists `route-rlp`, `section-rlp` and `gw-rlp` and  in the affection condition
* Verify `route-rlp` is **not** listed in the gateway or gateway listener status
